### PR TITLE
Precomp topsort

### PR DESCRIPTION
**Pull request by varun-sreedhar**
Pull Request for Precompilation

Features:
- Output to different strategy patterns
- Topologically sort tasks
- Parse .h files header comments

Structure of code:
- Main in RplTaskBuildMain.cpp
- Makefile for tests
- CMake to build tests
- PrecompilationCode folder contains all the code for sorting
- TestTaskFolder task header files
- RplTask.h/.cpp here for some reason idk why
- yeah

### CODE REVIEW

**Review by asbauer — APPROVED** [review]

File: PrecompilationCode/ArduinoStrategyPattern.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		task1->process();
+		task2->process();
+	}
+*/
+
+/**
+* Arduino Strategy pattern which when given a sorted list of tasks it
+* produces the code above and places the result in a created main.ino
+*/
+class ArduinoStrategyPattern : public IOutputStrategyPattern{
```
Comment (daftcube) [inline PrecompilationCode/ArduinoStrategyPattern.h:31]: We talked about this, but splitting this up into more digestable functions would be spicy.

File: PrecompilationCode/IOutputStrategyPattern.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#include <vector>
+#include <iostream>
+
+using namespace std;
+
+struct Task{
```
Comment (daftcube) [inline PrecompilationCode/IOutputStrategyPattern.h:6]: Preferably in its own file but apparently this is not trivial so don't sweat it.

File: PrecompilationCode/ArduinoStrategyPattern.h (hunk 1 above)
Comment (daftcube) [reply PrecompilationCode/ArduinoStrategyPattern.h:31]: This was fixed before I submitted my review. Looks great :)

**Review by daftcube — APPROVED** [review]
I approve

**Review by daftcube — APPROVED** [review]
LGTM

**Pull request merged**